src/nxt_array.h

```diff
@@ -9,34 +10,32 @@
 
 
 typedef struct {
-    void      *elts;
-    /* nelts has uint32_t type because it is used most often. */
-    uint32_t  nelts;
-    uint16_t  size;
-    uint16_t  nalloc;
-    nxt_mp_t  *mem_pool;
+    void                *elts;
+    nxt_uint_t          nelts;
+    nxt_uint_t          size;
+    nxt_uint_t          nalloc;
+    nxt_mp_t            *mem_pool;
 } nxt_array_t;
 
 
-nxt_inline void
-nxt_array_init(nxt_array_t *array, nxt_mp_t *mp, size_t size)
-{
-    array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));
```

Comment:
The pointer to "elts" was changing at the time of the memory reallocation. This offset would point to a non-existent pointer.

Reply:
It wouldn't, as by the time of array initialization there cannot be any reallocations. For already existing arrays the "nxt_array_reset()" function is supposed to be used.

Reply:
I meant memory reallocation in case of array expansion.

Reply:
"point to a non-existent pointer" makes no sense to me. Pointer is just an arbitrary integer value that is big enough to keep any possible memory address and can be used with pointer arithmetic. Depending on the purpose, it can hold a memory address that is valid for accessing (dereferencing), or not valid for accessing. Pointers to non-valid memory addresses are often used for pointer arithmetic or like some indicators.

E.g. there are can be pointers indicating the buffer end. In your terminology the "end" field of "nxt_buf_mem_t" structure always points "to a non-existent pointer" as it keeps the address of high border of the allocated buffer memory.

Here it points to the memory where elements supposed to start. If there's zero elements, then dereferencing the first element [0] of zero elements array is a mistake. And this mistake is not different from dereferencing of the second element [1] of 1 elements array, or dereferencing of the third element [2] of 2 elements array, and etc...

Reply:
I understand your logic of following the cursor outside the array. But it actually works until the first reallocation of memory (in the old code), since at that moment the memory is re-allocated in a separate block and a new pointer to a new memory area not related to the current array is assigned to array->elts and actually stops reflecting the real and actual boundary the end of the array.

Reply:
"stops reflecting the real and actual boundary the end of the array" and this indicates that memory isn't allocated with the array structure, but with additional nxt_mp_alloc() call and therefore can be freed in case of additional reallocation needed.

The nxt_array_init() function call reflects the same state as nxt_array_create() call with n set to 0.

Reply:
As you can see, you, like the programmer (about whom I wrote), also fell into this trap earlier (I speak with care). Especially in the context of how you advocate the importance of holding array bounds until the first reallocation of memory.

**Explanation:**
You say that `nxt_array_create` and `nxt_array_init` are essentially the same when `n = 0`, and they are. But for some reason you forget about what is expected after the execution of these functions, namely: `nxt_array_add`, which immediately resets the `array->elts` array end pointer to a new one. What then is the initial meaning of pointing to a neighboring object, if immediately after it loses this boundary? Taking into account what you wrote earlier about the need to specify the array boundary in case there are elements in the array, if boundary control may be necessary, then only in the case of `nxt_array_create` with a value of `n > 1` (not `n == 0`), since in any other case this the boundary is useless and ineffective and is essentially an over-indicator of anything that you are overly focused on and give unjustified importance to, like the programmer before you, which in my humble opinion is wrong and it confirms a number of irrefutable deductive facts.

Reply:
> I want to remind you that in the current version of the code (the old version of the code), you simply leave the initial allocated memory area abandoned, which is not normally freed for later reallocation of memory.

It will be freed when the request processing be finished. Keeping this structure even without using their fields simplifies the code, allowing to have no difference between cases when it was used and when it wasn't.

> What then is the initial meaning of pointing to a neighboring object, if immediately after it loses this boundary?

The initial meaning is to indicate that the `array->elts` pointer is a part of the array structure. It doesn't controls boundary anyhow. The allowed addresses are controlled by the `array->nelts` counter, where 0 means that there's no allowed elements for reading in by the `array->elts` pointer. When `array->nelts` set to 0, the `array->elts` pointer shouldn't be dereferenced and only indicates that there's no separate memory allocated for the elements.

> any other case this the boundary is useless and ineffective and is essentially an over-indicator of anything that you are overly focused on and give unjustified importance to, like the programmer before you, which in my humble opinion is wrong and it confirms a number of irrefutable deductive facts

If you're trying to say that the code here isn't easy, friendly and very intuitive, then you're right. Is it bad? It depends on the projects goals. One of the goals is to squeeze every bit of performance.

Reply:
> You have singled out only one use case out of several dozen different ones, using, among other things, the nxt_array_create you mentioned earlier, which, as you said, has the same state as nxt_array_init when n == 0.

Sure, it makes sense to talk only use cases existing in the current code. Sorry, I can't talk about any imaginable use-cases or future use-cases. It's not a public API, and we don't develop arrays for general purposes. This API focuses on the current code and current use-cases.

> For nxt_array_init, the memory for storing array->elts is generally initially is not allocated, so I don't understand what kind of freed memory you are talking about when mentioning requests.

The memory where the array structure itself is stored.

> Moreover nxt_array.c and nxt_array.h are not only used for queries and in other cases may have a longer abandoned and unreleased lifetime for array->elts using nxt_array_create. I remind you that we originally talked about wasted memory using nxt_array_create when n > 1.

Could you clarify what memory is wasted when nxt_array_create() is called with n > 1?

Please also take into account:
1. Saving on syscalls is a WAY more important from performance point of view than doing free as soon as possible.
2. All allocated memory from memory pools will be freed anyway with destruction of that pool.
3. In nginx, for example, you barely find any free() call in runtime. It only allocates memory and never frees it during request processing. So, in your terms it wastes a lot of memory, but in fact it just frees it with some delay (when the request processing ends). This delay isn't significant comparing to performance benefits of not freeing memory during request processing. Unit follows the same ideology, but due to more complex design of memory pools, it sometimes  calls nxt_mp_free().

> I don't understand where you saw that I suggest to dereference the pointer after it has been initialized? Please re-read my answer to you earlier: https://github.com/nginx/unit/pull/664#discussion_r836379187

I don't understand what is wrong with setting:
```
array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));
```
in context of `nxt_var_query_s`.

I don't care where it points in this case at all. It's the same as setting it to NULL, or 0... just some value. Then you can check for NULL, or can check for `nxt_pointer_to(array, sizeof(nxt_array_t));`. To avoid adding more conditions the value of `nxt_pointer_to(array, sizeof(nxt_array_t));` is used.

> Instead of being NULL for example until array->elts is fully initialized with nxt_array_add for example?

Then you will need additional conditions as instead of 2 states of this pointer:
1. point to the end of `nxt_array_t`
2. point to somewhere else

You will have 3 states of this flag:
1. NULL
2. point to the end of `nxt_array_t`
3. point to somewhere else

Then you will need to check for 3 states instead of 2 states.

And, if you will always allocate `array->elts` memory separate from the `nxt_array_t` structure, then you will always have 2 allocation instead of 1 allocation when creating a new array.

> As we have already been able to find out. Performance - in your case, just a cover for not considering a different point of view, forcing you to accept your own technically unreasonable.

Please provide a point that actually improves something and clear explanation of that improvement.

Check for 3 states instead of 2 states isn't improvement, but opposite - pessimization.
2 allocations instead of 1 allocation isn't improvement, but opposite - pessimization.